File: reverse_words.py
"""
Given a string of words, reverse all the words.
start = "This is the best."
finish = "best the is This."
"""

def reverse(s):
    word_list = s.split()
    word_list.reverse()
    
    return word_list
    
    
def reverse_2(s):
    return " ".join(reversed(s.split()))


def reverse_3(s):
    length = len(s)
    words = []
    chars_to_skip = [",", ";", "-", "?", "!", ".", " "]
    i = 0
    
    while i < length:
        if s[i] not in chars_to_skip:
            j = i
            word = ""
            while j < length and s[j] not in chars_to_skip:
                word += s[j]
                j += 1
            
            words.insert(0, word)
            i = j+1
        else:
            i += 1
            
    return words


def reverse_4(s):
    return " ".join(s.split()[::-1])
    
    
print(reverse("Hello there, how are you?"))
print(reverse_2("Hello there, how are you?"))
print(reverse_3("Hello  there ! ; ;;, how are you ?"))
print(reverse_4("Hello there, how are you?"))

File: ReverseWords.java
package reversewords;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ReverseWords {
    
    public String[] process1(String s) {
        String[] strArr = s.split(" ");
        
        for (int i = 0; i < strArr.length / 2; i++) {
            int lastPos = strArr.length - 1 - i;
            
            String temp = strArr[i];
            strArr[i] = strArr[lastPos];
            strArr[lastPos] = temp;
        }        
        
        return strArr;
    }
    
    public List<String> process2(String s) {
        List<String> strList = new ArrayList<>(Arrays.asList(s.split(" "))); 
        Collections.reverse(strList);
        
        return strList;
    }
    
    public List<String> process3(String s) {
        final List<Character> charsToRemove = Arrays.asList(new Character[] {'?', ';', ',', ':', ' '});
        
        StringBuilder word = new StringBuilder();
        List<String> result = new ArrayList<>();
        int i = 0;
        while (i < s.length()) {
            word.setLength(0);
            int j = i;
            while (j < s.length() && !charsToRemove.contains(s.charAt(j))) {
                word.append(s.charAt(j));
                j++;
            }
            result.add(0, word.toString());
            i = j+1;            
        }
        
        return result;        
    }
    
    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        
        ReverseWords app = new ReverseWords();
        
        String str = "Hello there, how are you";
               
        System.out.println(app.process2(str));
        System.out.println(app.process3(str));        
    }
    
}
